src/ucp/core/ucp_proxy_ep.c

```diff
@@ -0,0 +1,209 @@
+    UCP_PROXY_EP_SET_OP(ep_am_bcopy);
+    UCP_PROXY_EP_SET_OP(ep_am_zcopy);
+    UCP_PROXY_EP_SET_OP(ep_atomic_add64);
+    UCP_PROXY_EP_SET_OP(ep_atomic_fadd64);
+    UCP_PROXY_EP_SET_OP(ep_atomic_swap64);
+    UCP_PROXY_EP_SET_OP(ep_atomic_cswap64);
+    UCP_PROXY_EP_SET_OP(ep_atomic_add32);
+    UCP_PROXY_EP_SET_OP(ep_atomic_fadd32);
+    UCP_PROXY_EP_SET_OP(ep_atomic_swap32);
+    UCP_PROXY_EP_SET_OP(ep_atomic_cswap32);
+    UCP_PROXY_EP_SET_OP(ep_tag_eager_short);
+    UCP_PROXY_EP_SET_OP(ep_tag_eager_bcopy);
+    UCP_PROXY_EP_SET_OP(ep_tag_eager_zcopy);
+    UCP_PROXY_EP_SET_OP(ep_tag_rndv_zcopy);
+    UCP_PROXY_EP_SET_OP(ep_tag_rndv_cancel);
+    UCP_PROXY_EP_SET_OP(ep_tag_rndv_request);
+    UCP_PROXY_EP_SET_OP(ep_pending_add);
+    UCP_PROXY_EP_SET_OP(ep_pending_purge);
+    UCP_PROXY_EP_SET_OP(ep_flush);
+    UCP_PROXY_EP_SET_OP(ep_fence);
+    UCP_PROXY_EP_SET_OP(ep_check);
+    UCP_PROXY_EP_SET_OP(ep_destroy);
+    UCP_PROXY_EP_SET_OP(ep_get_address);
+    UCP_PROXY_EP_SET_OP(ep_connect_to_ep);
+
+    self->iface.ops.iface_tag_recv_zcopy     = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_tag_recv_cancel    = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.ep_create                = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.ep_create_connected      = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_flush              = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_fence              = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_progress_enable    = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_progress_disable   = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_progress           = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_event_fd_get       = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_event_arm          = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_close              = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_query              = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_get_device_address = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_get_address        = (void*)ucp_proxy_ep_fatal;
+    self->iface.ops.iface_is_reachable       = (void*)ucp_proxy_ep_fatal;
+
+    return UCS_OK;
+}
+
+static UCS_CLASS_CLEANUP_FUNC(ucp_proxy_ep_t)
+{
+    if ((self->uct_ep != NULL) && self->is_owner) {
+        uct_ep_destroy(self->uct_ep);
+    }
+}
+
+void ucp_proxy_ep_replace(ucp_proxy_ep_t *proxy_ep)
+{
+    ucp_ep_h ucp_ep = proxy_ep->ucp_ep;
+    ucp_lane_index_t lane;
+
+    ucs_assert(proxy_ep->uct_ep != NULL);
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        if (ucp_ep->uct_eps[lane] == &proxy_ep->super) {
+            ucp_ep->uct_eps[lane] = proxy_ep->uct_ep;
+            proxy_ep->uct_ep = NULL;
+            break;
+        }
+    }
+
+    uct_ep_destroy(&proxy_ep->super);
+}
+
+void ucp_proxy_ep_set_uct_ep(ucp_proxy_ep_t *proxy_ep, uct_ep_h uct_ep,
+                             int is_owner)
```

Comment:
in next pull request which would introduce proxy_ep for sending signaled active message